qiskit_experiments/framework/settings.py

```diff
@@ -0,0 +1,102 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Settings mixing class
+"""
+
+import inspect
+from collections import OrderedDict
+from functools import wraps
+from typing import Dict, Any
+
+
+class Settings:
+    """Class mixing for storing instance init settings.
+
+    This mixin adds a ``__new__`` method that stores the values of args
+    and kwargs passed to the class instances ``__init__`` method and a
+    ``settings`` property that returns an ordered dict of these values.
+
+    .. note::
+
+        This mixin is intended as a mixing for base classes so that when
+        creating subclasses, those subclasses can inherit the logic for
+        saving and returning settings.
+
+        Note that there is small performance overhead to initializing classes
+        with this mixin so it should not be used for adding settings to all
+        classes without consideration. For classes that already store values
+        required to recover the ``__init__`` args they should instead
+        implement an appropriate :meth:`settings` property directly.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        # This method automatically stores all arg and kwargs from subclass
+        # init methods for use in converting an experiment to config
+
+        # Get all non-self init args and kwarg names for subclass
+        spec = inspect.getfullargspec(cls.__init__)
+        init_arg_names = spec.args[1:]
+        num_init_kwargs = len(spec.defaults) if spec.defaults else 0
+        num_init_args = len(init_arg_names) - num_init_kwargs
+
+        # Convert passed values for args and kwargs into an ordered dict
+        # This will sort args passed as kwargs and kwargs passed as
+        # positional args in the function call
+        num_call_args = len(args)
+        ord_args = OrderedDict()
+        ord_kwargs = OrderedDict()
+        for i, argname in enumerate(init_arg_names):
```

Comment:
Seems like this doesn't assume variadic arguments in the constructor. 

https://github.com/Qiskit/qiskit-experiments/blob/1c3ebe3ea439613d4b57aaa63ee36af7d0a0f703/qiskit_experiments/library/characterization/cr_hamiltonian.py#L129-L136
for example, `CrossResonanceHamiltonian` experiment has `**kwargs` for single liner execution (if we explicitly write every configurable pulse parameters the constructor signature will be very long). Of course user can update parameters from `set_experiment_options`.

Reply:
Good catch that it doesnt work with generic `*args **kwargs`. One thing I wanted to do was make sure it only stores non-default kwargs (so kwargs actually passed to the init function). And correctly extract positional args passed as kwargs into the stored args, and kwargs passed as positional args into the stored kwargs. This is why I didn't originally go with the signature approach you suggest. I'll try playing around with this some more to see if I can come up with something that works for both cases.